.github/CONTRIBUTING.md

```diff
@@ -1,8 +1,65 @@
-# How to Contribute
-
 We'd love to accept your patches and contributions to this project. There are
 just a few small guidelines you need to follow.
 
+## Repository Structure
+
+There are three main sections:
+
+**Codelabs** are interactive tutorials that are written in markdown syntax and published at [blocklycodelabs.dev](https://blocklycodelabs.dev). Codelabs mix natural language, code samples, and screenshots. The target user is following along and running the code as they read.
+
+The codelabs directory has a [template](https://github.com/google/blockly-samples/blob/master/codelabs/template.md) and one folder per codelab. Each codelab's folder contains the codelab (as a markdown file), and all associated assets (pngs, gif, etc).
+
+**Examples** are sample projects showing how to use Blockly features. Examples contain mostly code and a demo web page. Example code should be extremely well-commented to make it easy to copy. The target user may be reading the code, running it locally, or copying code snippets.
+
+The examples directory has one folder per example. Each example can be run with `npm install && npm run start`, and has a `README.md` file with additional context or instructions.
+
+**Plugins** are self-contained pieces of code that add functionality to Blockly. Plugins can add fields, define themes, create renderers, and much more. The target user is a developer who finds and uses the plugin through npm. Plugins defined in this repository are *first-party* plugins, which means that they are supported by the Blockly team.
+
+The plugins directory has one folder per plugin. To create the correct directory structure for your plugin, use [@blockly/create-package](https://github.com/google/blockly-samples/tree/master/plugins/dev-create#blocklycreate-package-).
+
+## How can I contribute?
+
+- Suggest plugins, examples, or codelabs
+- Implement plugins, examples, or codelabs
+- Report bugs
+- Fix bugs
+- Reproduce bugs
+
+## How do I add a plugin?
+
+Plugins go through four stages: suggestion, discussion, implementation, and publishing:
```

Comment:
I'm just going to be candid with you, I have a few misgivings about the suggestion -> discussion -> implementation process.

Imagine that I have a brilliant idea for a plugin. I get super excited and I write a really detailed feature suggestion. I'm thinking about it all the time, and I just /know/ that you guys are going to accept it. But then... you don't. Maybe the idea isn't as brilliant as I thought it was. Or maybe it just isn't a good fit for a first party plugin. But either way it's pretty demotivating. It can feel like not getting accepted means the idea is bad (even if it absolutely doesn't!).

Can you guys think of any changes that would help avoid this? Maybe just a simple flip from "seeking approval" to to something else would help.

Reply:
Would it help if I had a section or page on criteria for inclusion as a first-party plugin?

I don't see a way to get around the disappointment of having something not accepted as a first-party plugin, but I'm writing an explicit process here so that a contributor knows what to expect, and knows that the first step is not to make the complete PR--which I think would be even more crushing.

Reply:
I just realized you might be asking if there's a clear process to move from the suggestion phase to the design phase.  I had assumed that it moved immediately upon making a feature request, but I can see that it would be good to get a preliminary yes/no from the team.  I'll think on that too.

Reply:
> Would it help if I had a section or page on criteria for inclusion as a first-party plugin?

I really like this idea! I think giving people an idea of what you're looking for would really help. Then you could easily tell people "yeah we think your idea is really cool, and you should definitely still do it! we just think it would work better in it's own repo, rather than including it in this one. we like to make sure plugins in this repo are x y z and do a b c as detailed at <see page here>"

It still encourages them to go and build stuff, but you can make sure you're only first-party-supporting plugins you want to.

Reply:
Added a note "explain why it should be a first-party plugin rather than a third-party plugin".  Still need to write up and publish explicit criteria.